### PR TITLE
feat: Add autodiff driven SciPy optimizer

### PR DESCRIPTION
**Pull request by lukasheinrich**
# Description

This adds a optimizer based on scipy. The objective and gradient are computed by the respective backend but the parameter stepping is done by `scipy.minimize` since #583 seems to be blocked by https://github.com/tensorflow/probability/issues/637 and additionally it's not clear we can implement parameter bounds (which in scipy need `SLSQP` method)

this optimizer is subclassable by any autodifferentiable backend.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Drop tensor-ful hand-written optimization for scipy.minimize
* Leverage autodiff capability in TensorFlow and PyTorch to pass exact gradients/jacobians
* Can reproduce sbottom contour w/ TF and PyTorch (PyTorch is faster than NumPy)
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
@lukasheinrich Is this ready for review, or is there further WIP?

**Comment by lukasheinrich** [comment]
yeah this is ready for review

**Comment by lukasheinrich** [comment]
this is ready for review @kratsg @matthewfeickert

File: src/pyhf/optimize/opt_tflow.py [hunk 1]
```diff
@@ -1,76 +1,55 @@
...
+        data = tf.placeholder(tf.float32, (m.config.nmaindata + m.config.nauxdata,))
+        nll = objective(pars, data, m)
+        nllgrad = tf.identity(tf.gradients(nll, pars)[0])
+        func = eval_func([nll, nllgrad], pars, data, d)
+        return func, init_pars, par_bounds
+
+    def setup_constrained(self, objective, poival, m, d, init_pars, par_bounds):
+        tensorlib, _ = get_backend()
+        data = tf.placeholder(tf.float32, (m.config.nmaindata + m.config.nauxdata,))
+        idx = default_backend.astensor(range(m.config.npars), dtype='int')
```
Comment (kratsg) [inline src/pyhf/optimize/opt_tflow.py]: probably should support `tensorlib.range`

File: src/pyhf/pdf.py [hunk 2]
```diff
@@ -422,6 +422,7 @@ def __init__(self, spec, batch_size=None, **config_kwargs):
             if hasattr(parset, 'pdf_type'):  # is constrained
                 self.config.auxdata += parset.auxdata
                 self.config.auxdata_order.append(k)
+        self.config.nauxdata = len(self.config.auxdata)
```
Comment (kratsg) [inline src/pyhf/pdf.py]: why does this get set later instead of at `init` time?

File: src/pyhf/optimize/opt_tflow.py (hunk 1 above)
Comment (lukasheinrich) [reply src/pyhf/optimize/opt_tflow.py]: PRs welcome :)

File: src/pyhf/pdf.py (hunk 2 above)
Comment (lukasheinrich) [reply src/pyhf/pdf.py]: because the stuff that `appends` to auxdata is done here see comment above

```
        # this is tricky, must happen before constraint
        # terms try to access auxdata but after
        # combined mods have been created that
        # set the aux data
```

File: src/pyhf/optimize/autodiff.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        )
+        unconstr_pars = fitresult.x
+        tensorlib, _ = get_backend()
+        return tensorlib.astensor(unconstr_pars)
+
+    def constrained_bestfit(
+        self, objective, constrained_mu, data, pdf, init_pars, par_bounds
+    ):
+        poival = constrained_mu
+        for_constr = self.setup_constrained(
```
[the remark this answers is not in the document]
Comment (lukasheinrich) [reply src/pyhf/optimize/autodiff.py]: done

File: src/pyhf/optimize/autodiff.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import scipy
+from .. import default_backend
+from .. import get_backend
+
+
+class AutoDiffOptimizerMixin(object):
+    def unconstrained_bestfit(self, objective, data, pdf, init_pars, par_bounds):
+        for_unconstr = self.setup_unconstrained(
```
[the remark this answers is not in the document]
Comment (lukasheinrich) [reply src/pyhf/optimize/autodiff.py]: done

**Comment by lukasheinrich** [comment]
pushed some more commits. PTAL

File: src/pyhf/pdf.py (hunk 2 above)
Comment (kratsg) [reply src/pyhf/pdf.py]: i see comment now. yes. this makes sense... not obvious initially.

File: src/pyhf/optimize/opt_tflow.py (hunk 1 above)
Comment (kratsg) [reply src/pyhf/optimize/opt_tflow.py]: #636 is filed for now.

File: src/pyhf/optimize/opt_pytorch.py [hunk 5]
```diff
@@ -1,58 +1,46 @@
...
+
+
+class pytorch_optimizer(AutoDiffOptimizerMixin):
+    def __init__(self, *args, **kargs):
+        pass
+
+    def setup_unconstrained(self, objective, pdf, data, init_pars, par_bounds):
+        def func(pars):
+            tensorlib, _ = get_backend()
+            pars = tensorlib.astensor(pars)
```
Comment (kratsg) [inline src/pyhf/optimize/opt_pytorch.py]: why not `pars = tensorlib.astensor(pars, requires_grad=True)`?

File: src/pyhf/optimize/opt_pytorch.py [hunk 6]
```diff
@@ -1,58 +1,46 @@
...
+            pars.requires_grad = True
+            r = objective(pars, data, pdf)
+            grad = torch.autograd.grad(r, pars)[0]
```
Comment (kratsg) [inline src/pyhf/optimize/opt_pytorch.py]: what's the `[0]` for? do we throw away the rest? Why not

`grad, _ = torch.autograd.grad(r, pars)`

Comment (kratsg) [reply src/pyhf/optimize/opt_pytorch.py]:
```suggestion
            grad, _ = torch.autograd.grad(r, pars)
```

File: src/pyhf/optimize/opt_pytorch.py (hunk 5 above)
Comment (kratsg) [reply src/pyhf/optimize/opt_pytorch.py]:
```suggestion
            pars = tensorlib.astensor(pars, requires_grad=True)
```

File: src/pyhf/optimize/opt_pytorch.py [hunk 7]
```diff
@@ -1,58 +1,46 @@
...
+            return r.detach().numpy(), grad
```
Comment (kratsg) [inline src/pyhf/optimize/opt_pytorch.py]: why detach?

File: src/pyhf/optimize/opt_pytorch.py (hunk 5 above)
Comment (matthewfeickert) [reply src/pyhf/optimize/opt_pytorch.py]: Just to check, the only reason that we need to explicitly cast as a tensor here is because these `pars` are being passed into `func` by the internals of

```
fitresult = scipy.optimize.minimize(func, init, method='SLSQP', jac=True, bounds=bounds)
```

which won't be using tensors, correct? We can't avoid having to do this cast?

File: src/pyhf/optimize/opt_pytorch.py [hunk 8]
```diff
@@ -1,58 +1,46 @@
...
+
+        return func, init_pars, par_bounds
+
+    def setup_constrained(self, objective, poival, pdf, data, init_pars, par_bounds):
+        tensorlib, _ = get_backend()
+        idx = default_backend.astensor(range(pdf.config.npars), dtype='int')
+        init = default_backend.astensor(init_pars)
```
Comment (matthewfeickert) [inline src/pyhf/optimize/opt_pytorch.py]: See question above RE: if the `astensor` calls that don't explicitly transform the `dtype` are needed as a result of use in `scipy.optimize.minimize`.

File: src/pyhf/optimize/opt_pytorch.py [hunk 9]
```diff
@@ -1,58 +1,46 @@
...
+        nuisinit = default_backend.concatenate(
+            [init[: pdf.config.poi_index], init[pdf.config.poi_index + 1 :]]
+        ).tolist()
+        nuisidx = default_backend.concatenate(
+            [idx[: pdf.config.poi_index], idx[pdf.config.poi_index + 1 :]]
+        ).tolist()
+        nuisbounds = [par_bounds[i] for i in nuisidx]
+        tv = _TensorViewer([[pdf.config.poi_index], nuisidx])
+
+        data = tensorlib.astensor(data)
```
Comment (matthewfeickert) [inline src/pyhf/optimize/opt_pytorch.py]: See question above RE: if the `astensor` calls that don't explicitly transform the `dtype` are needed as a result of use in `scipy.optimize.minimize`.

File: src/pyhf/optimize/opt_pytorch.py [hunk 10]
```diff
@@ -1,58 +1,46 @@
...
+        poivals = tensorlib.astensor([poival], dtype='float')
+
+        def func(pars):
+            pars = tensorlib.astensor(pars)
```
Comment (matthewfeickert) [inline src/pyhf/optimize/opt_pytorch.py]: See question above RE: if the `astensor` calls that don't explicitly transform the `dtype` are needed as a result of use in `scipy.optimize.minimize`.

File: src/pyhf/optimize/opt_tflow.py [hunk 11]
```diff
@@ -1,76 +1,64 @@
...
+        data_placeholder = tf.placeholder(
+            tf.float32, (pdf.config.nmaindata + pdf.config.nauxdata,)
         )
-        objective = objective(pars, data, pdf)
-        hessian = tf.hessians(objective, nuis_cat)[0]
-        gradient = tf.gradients(objective, nuis_cat)[0]
-        invhess = tf.linalg.inv(hessian)
-        update = tf.transpose(tf.matmul(invhess, tf.transpose(tf.stack([gradient]))))[0]
+        idx = default_backend.astensor(range(pdf.config.npars), dtype='int')
+        init = default_backend.astensor(init_pars)
```
Comment (matthewfeickert) [inline src/pyhf/optimize/opt_tflow.py]: See question above RE: if the `astensor` calls that don't explicitly transform the `dtype` are needed as a result of use in `scipy.optimize.minimize`.

File: src/pyhf/optimize/opt_pytorch.py (hunk 5 above)
Comment (lukasheinrich) [reply src/pyhf/optimize/opt_pytorch.py]: because scipy doesn't know anything about tensors other than numpy, so we just use the backedns to compute the logpdf -> cast to numpy, and scipy.optimize will return a new parameter vector (as numpy array) at the location where it wants to evaluate a likelihood nexxt

File: src/pyhf/optimize/autodiff.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
+import scipy
+from .. import default_backend
+from .. import get_backend
+
+
+class AutoDiffOptimizerMixin(object):
+    def unconstrained_bestfit(self, objective, data, pdf, init_pars, par_bounds):
+        func, init, bounds = self.setup_unconstrained(
+            objective, pdf, data, init_pars, par_bounds
```
[the remark this answers is not in the document]
Comment (lukasheinrich) [reply src/pyhf/optimize/autodiff.py]: done

File: src/pyhf/optimize/opt_pytorch.py (hunk 6 above)
Comment (lukasheinrich) [reply src/pyhf/optimize/opt_pytorch.py]: it's not  tuple return, but a one-element list.. i'd chose `a,_` only for tuple returns tbh

File: src/pyhf/optimize/opt_pytorch.py (hunk 7 above)
Comment (lukasheinrich) [reply src/pyhf/optimize/opt_pytorch.py]: because it needs to be detached from the gradient calculation in  pytorch before it can be cast to numpy

File: src/pyhf/optimize/opt_tflow.py [hunk 13]
```diff
@@ -1,76 +1,55 @@
...
-                nuis_cat[: pdf.config.poi_index],
-                poi_par,
-                nuis_cat[pdf.config.poi_index :],
-            ]
+def eval_func(op, argop, dataop, d):
+    def func(p):
+        tensorlib, _ = get_backend()
+        xx = tensorlib.tolist(p) if isinstance(p, tf.Tensor) else p
+        yy = tensorlib.tolist(d) if isinstance(d, tf.Tensor) else d
+        value = tensorlib.session.run(op, feed_dict={argop: xx, dataop: yy})
```
[the remark this answers is not in the document]
Comment (lukasheinrich) [reply src/pyhf/optimize/opt_tflow.py]: done

File: src/pyhf/optimize/opt_pytorch.py [hunk 14]
```diff
@@ -1,58 +1,46 @@
...
+    def __init__(self, *args, **kargs):
+        pass
+
+    def setup_unconstrained(self, objective, data, pdf, init_pars, par_bounds):
+        def func(pars):
+            tensorlib, _ = get_backend()
+            pars = tensorlib.astensor(pars)
+            pars.requires_grad = True
+            r = objective(pars, data, pdf)
+            grad, _ = torch.autograd.grad(r, pars)
```
Comment (kratsg) [inline src/pyhf/optimize/opt_pytorch.py]:
```suggestion
            grad = torch.autograd.grad(r, pars)[0]
```
